SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,950 @@
+
+        // Write one sector at a time with the metadata attached
+        local obj_pos = 0;
+        local obj_remaining = obj_len;
+        do {
+            local written = _sectorsMap[_at_sec].write(obj, obj_pos, obj_remaining);
+            obj_remaining = obj_remaining - written;
+            obj_pos += written;
+            // Object (or partial object) should allocate all memory on the flash sector
+            // otherwise we will have gap in object writing.
+            if (obj_remaining != 0 && _sectorsMap[_at_sec].getFreeSpace() > 0)
+                throw "Failed to write data on the flash.";
+
+            if (_sectorsMap[_at_sec].getFreeSpace() <= SPIFLASHLOGGER_OBJECT_MIN_SIZE) {
+                _sectorsMap[_at_sec].closeForWriting();
+                _at_sec = (_at_sec + 1) % _sectors;
+                // erase and mark as start next sector
+                _sectorsMap[_at_sec].prepareForWrite();
+                _at_pos = _sectorsMap[_at_sec].getWritePosition();
+            } else {
+                _at_pos += written;
+            }
+        } while (obj_remaining > 0);
+
+        _disable();
+    }
+
+    function read(onData = null, onFinish = null, step = 1, skip = 0) {
+        assert(typeof step == "integer" && step != 0);
+
+        if (skip < 0) {
+            throw "Wrong skip parameter value.";
+        }
+
+        // skipped tracks how many entries we have skipped, in order to implement skip
+        local skipped = math.abs(step) - skip - 1;
+        local count = 0;
+        local objectIterator = _getObjectIterator(step);
+        // There is no start point for object iteration
+        if (objectIterator == null) {
+            if (onFinish != null)
+                onFinish();
+            return;
+        }
+
+        local continueCallback, readNextObject;
+        // Passed in to the read callback to be called as `next`
+        continueCallback = function(keepGoing = true) {
+            if (keepGoing == false) {
+                // Clean up and exit
+                objectIterator = null;
+                if (onFinish != null)
+                    onFinish();
+            } else {
+                // There are more objects to read, read the next one
+                return imp.wakeup(0, readNextObject.bindenv(this));
+            }
+        };
+
+        readNextObject = function() {
+            local obj = objectIterator.getNextObject(step, skip);
+            if (obj != null && obj.isValid()) {
+                return onData(obj.getPayload(),
+                    obj.getAddress(), continueCallback.bindenv(this));
+            } else {
+                return continueCallback(false);
+            }
+            return true;
+        }.bindenv(this);
+
+        // start reading objects in this sector
+        imp.wakeup(0, function() {
+            return continueCallback(true);
+        }.bindenv(this));
+    }
+
+    function readSync(index) {
+        //  index == 0 corresponding write postion
```

Comment:
what does this comment mean?

Reply:
changed comment, but it means that 0 - is equal to the current write position and we have no object at the current write position.